spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraPartitionKeyRDD.scala

```diff
@@ -0,0 +1,125 @@
+
+// O[ld] Is the type of the RDD we are Mapping From, N[ew] the type were are mapping too Old
+class CassandraPartitionKeyRDD[O, N] private[connector] (prev: RDD[O],
+                                      keyspaceName: String,
+                                      tableName: String,
+                                      connector: CassandraConnector,
+                                      columns: ColumnSelector = AllColumns,
+                                      where: CqlWhereClause = CqlWhereClause.empty,
+                                      readConf: ReadConf = ReadConf())
+                                      (implicit oldTag: ClassTag[O], newTag: ClassTag[N],
+                                       @transient rwf: RowWriterFactory[O], @transient rrf: RowReaderFactory[N])
+  extends CassandraRDD[N](prev.sparkContext, connector, keyspaceName, tableName, columns, where, readConf, prev.dependencies) {
+
+  //Make sure copy operations make new CPKRDDs and not CRDDs
+  override def copy(columnNames: ColumnSelector = columnNames,
+                   where: CqlWhereClause = where,
+                   readConf: ReadConf = readConf, connector: CassandraConnector = connector): CassandraPartitionKeyRDD[O,N] =
+    new CassandraPartitionKeyRDD[O,N](prev, keyspaceName, tableName, connector, columnNames, where, readConf)
+
+  private val converter = ReplicaMapper[O](connector, keyspaceName, tableName)
+
+  //We need to make sure we get selectedColumnNames before serialization so that our RowReader is
+  //built
+  private val singleKeyCqlQuery: (String) = {
+    logDebug("Generating Single Key Query Prepared Statement String")
+    val columns = selectedColumnNames.map(_.cql).mkString(", ")
+    val partitionWhere = tableDef.partitionKey.map(_.columnName).map(name => s"$name = :$name")
+    val filter = (where.predicates ++ partitionWhere).mkString(" AND ")
+    val quotedKeyspaceName = quote(keyspaceName)
+    val quotedTableName = quote(tableName)
+    logDebug(s"SELECT $columns FROM $quotedKeyspaceName.$quotedTableName WHERE $filter")
+    (s"SELECT $columns FROM $quotedKeyspaceName.$quotedTableName WHERE $filter")
+  }
+
+  private def keyByReplica(implicit rwf: RowWriterFactory[O]): RDD[(Set[InetAddress], O)] = {
+    prev.mapPartitions( primaryKey =>
+      converter.keyByReplicas(primaryKey)
+    )
+  }
+
+
+  /**
+   * When computing a CassandraPartitionKeyRDD the data is selected via single CQL statements
+   * from the specified C* Keyspace and Table. This will be preformed on whatever data is
+   * avaliable in the previous RDD in the chain.
+   * @param split
+   * @param context
+   * @return
+   */
+  override def compute(split: Partition, context: TaskContext): Iterator[N] = {
+    connector.withSessionDo { session =>
+      logDebug(s"Query::: $singleKeyCqlQuery")
+      val stmt = session.prepare(singleKeyCqlQuery).setConsistencyLevel(consistencyLevel)
+      fetchIterator(session, stmt, prev.iterator(split, context))
```

Comment:
It looked like the session isn't closed until the iterator is exhausted in CassandraRDD. Maybe i'm misreading?